apps/src/music/player/interfaces/PlaybackEvent.ts

```diff
@@ -23,4 +23,6 @@ export interface PlaybackEvent {
   blockId: string;
   /** A unique ID used to group same sounds together in the timeline */
   id: string;
+  /** A list of contextual block types associated with a block's parents */
+  parentControlTypes: string[];
```

Comment:
Hmm...I don't love that we have to track this data for every generated playback event, but also can see why we don't want to do a lookup based on the block ID for every event on every validation pass...could we precompute these validation conditions ahead of time every time the workspace updates (using the blockspace change hook) and store these in a singleton? I think we were talking about a similar approach for other blockly-related validation conditions right?

Reply:
I'll look into this some more. The biggest challenge might be in confidently determining that something a repeat block was used _correctly_ just by analyzing the workspace blocks. Functions and triggers make this more complicated. By adding this data to the playback events, we know that the sound has played.

For example, to know if a `play` block is used with a loop correctly, we'd not only need to check that the blocks are nested, but also confirm  they aren't disabled/orpahned on the workspace. There also needs to be a valid top block. If it's under `when run`, we can assume it's right, but if the top block is a function definition, we'd also need to check that there's a call block for it elsewhere.

Reply:
I explored this some more here: https://github.com/code-dot-org/code-dot-org/compare/mike/blockly-condition-tracker

Unfortunately, I think it's impossible to determine things like the number of loop iterations just by looking at the Blockly workspace.

Reply:
Thanks for looking into this! Yeah, as discussed offline, your approach seems to be the best given the challenges with inspecting the blockly workspace directly.